### DISCUSSION TITLE
Impl `AsExpression` and `FromSql` for NonZero*

### DISCUSSION THREAD

**Discussion by Elrendio** [diesel-rs/diesel#3431] [Ideas]
# List of types
- [NonZeroI8](https://doc.rust-lang.org/core/num/struct.NonZeroI8.html)
- [NonZeroI16](https://doc.rust-lang.org/core/num/struct.NonZeroI16.html)
- [NonZeroI32](https://doc.rust-lang.org/core/num/struct.NonZeroI32.html)
- [NonZeroI64](https://doc.rust-lang.org/core/num/struct.NonZeroI64.html)
- [NonZeroI128](https://doc.rust-lang.org/core/num/struct.NonZeroI128.html)
- [NonZeroIsize](https://doc.rust-lang.org/core/num/struct.NonZeroIsize.html)
- [NonZeroU8](https://doc.rust-lang.org/core/num/struct.NonZeroU8.html)
- [NonZeroU16](https://doc.rust-lang.org/core/num/struct.NonZeroU16.html)
- [NonZeroU32](https://doc.rust-lang.org/core/num/struct.NonZeroU32.html)
- [NonZeroU64](https://doc.rust-lang.org/core/num/struct.NonZeroU64.html)
- [NonZeroU128](https://doc.rust-lang.org/core/num/struct.NonZeroU128.html)
- [NonZeroUsize](https://doc.rust-lang.org/core/num/struct.NonZeroUsize.html)

If you validate we (Stockly) can provide a PR 😊

**Comment by weiznich** [comment]
It really depends on the proposed mapping. Therefore I cannot answer that in general for all types.

These types do currently not have any matching diesel mapping for there non-`NonZero*` variants:

* NonZeroI128
* NonZeroIsize
* NonZeroU128
* NonZeroUsize

which essentially means: We will likely not want to support them at all.

For the other types it really depends on answers to the following questions:

* Which mappings to what SQL types are proposed?
* Which backends should be supported for which types? (Unsigned variants should be likely mysql specific)
* As all of the SQL databases allow to store a zero value in columns of the corresponding fields: How do we handle getting back a zero value from the database? Only support mappings to `Option<NonZero*>`? Something else? (The general rule here is to avoid runtime errors)
